clients/src/main/java/org/apache/kafka/clients/admin/DescribeUserScramCredentialsResult.java

```diff
@@ -0,0 +1,50 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kafka.clients.admin;
+
+import org.apache.kafka.common.KafkaFuture;
+import org.apache.kafka.common.annotation.InterfaceStability;
+
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ * The result of the {@link Admin#describeUserScramCredentials()} call.
+ *
+ * The API of this class is evolving, see {@link Admin} for details.
+ */
+@InterfaceStability.Evolving
+public class DescribeUserScramCredentialsResult {
+    private final KafkaFuture<Map<String, UserScramCredentialsDescription>> future;
```

Comment:
If we're going to have a per-user error, then we need `Map<String, KafkaFuture<...>>`.  This will also be useful if we need to add more per-user errors in the future.

Reply:
Now that I'm working on this, I discovered that there is no other API that can describe or list everything that works this way.  Everything that can describe or list everything returns a single future.  Every describe or list API that returns a map of keys to futures requires a non-empty list of keys to describe or list.  For example:

1. `listTopics()` lists all topics and returns a single `Future`; the `describeTopics()` API returns a map of names to futures but requires a non-empty list of topics to describe.
2. `describeConfigs()` returns a map of resources to futures but requires a non-empty list of resources to describe.
3. `describeLogDirs()` returns a map of broker IDs to futures but requires a non-empty list of brokers to describe.
4. `describeReplicaLogDirs()` returns a map of replicas to futures but requires a non-empty list of replicas to describe.
5. `describeConsumerGroups()` returns a map of consumer groups to futures but requires a non-empty list of consumer groups to describe.
6. `listPartitionReassignments()` allows listing all or a subset of reassignments and returns a single future.
7. `listOffsets()` returns a map of topic-partitions to futures but requires a non-empty list of topic-partitions to describe.
8. `describeClientQuotas()` allows listing all or a subset of quotas and returns a single future.

I think if we made this change here we would be off the beaten path.  That's not necessarily bad, but what tipped me off to this was the fact that when we list everything we have to create a future for every user that gets returned, and we don't know that list of users when we make the request, so there's really no way to implement it.

We could create two separate APIs: one for describing some explicit, non-empty list of users, which would return a map of users to futures, and another one that describes everything, which returns a single future.  `listTopics()` vs `describeTopics()` works this way, for example, though the information returned in the two is very different: when listing you just get the names, and when describing you get a lot more.  I don't see us distinguishing between listing vs. describing in terms of data -- we are going to send back the same two things (mechanism and iterations) regardless.  So we would probably be talking about creating a `describeUserScramCredentials()` API and a `describeAllUserScramCredentials()` API with the first taking a list and returning a map of futures and the second not taking a list and returning a single future.

But I'm thinking we should just keep it the way it is -- take a possibly-empty list and return a single future regardles of whether the list was empty or not.

Thoughts?

Reply:
I don't see the need for multiple futures; everything is returning in a single response from Kafka, and I can't think of why it would be necessary to have some users' credential descriptions available to the API caller but others not.  I'll take a look at keeping a single future for the response but moving the error status from the top level down into each per-user result.

Reply:
I kept the top-level error information but added per-user error information in addition.

Reply:
Thinking about this more, I would prefer having both a list RPC to show everything, and a describe RPC to show only some things.  There are a few reasons why.  One is that even though we currently only make one RPC, in the future we might make more than one.  In that case we would want multiple futures.

Another is that the general pattern in Kafka is that list RPCs show everything, but describe RPCs show only some things.  It's true that there are some places where we violate this pattern, but I still think it's worth trying to follow where we can.  Maybe this should be documented better so that when we add new RPCs, people aren't confused about whether to use "list" or "describe."

I also feel like in AdminClient, errors should be handled with futures pretty much all the time, unless there is a really strong reason not to.  This allows people to use an async style of programming.  In contrast, mixing in some errors that aren't futures, but need to be checked explicitly is very likely to confuse people.

Reply:
> a list RPC to show everything, and a describe RPC to show only some things.

Do you mean a list RPC that takes no arguments and returns every credential defined for every users and a describe RPC that takes 1 or more users and returns every credential defined for those specified users, and they both return the same information for each credential?

Or do you mean a list RPC and a describe RPC that return different sets of information (as is done with list vs. describe topics)?  I think you mean the former (two RPCs, each returning the same thing), but I want to be certain I understand.

> There are a few reasons why. One is that even though we currently only make one RPC, in the future we might make more than one. In that case we would want multiple futures.

I don't understand what this is referring to.  By "we currently only make one RPC" to what are you referring?

> I also feel like in AdminClient, errors should be handled with futures pretty much all the time

Agreed.  Will convert to using futures always, whenever we arrive at the final decision on what the RPCs need to look like.

I'm wondering if we convert to returning futures everywhere, can we stick with the one describe RPC?  For example, could the admin client return a `Future<Map<String, Future<Void>>>`?  Would that work, and if so, would that be a reasonable way to proceed?

Reply:
If we do decide to go with 2 separate APIs, then I might be concerned about using `list()` vs `describe()` if they return the same set of information (i.e. mechanism and iterations).  Perhaps using two separate names gives us room to expand `describe()` to return more information later on, though.  But if not, and they will always return the same information, then maybe `describeAll()` and `describe()` (or `listAll()` and `list()`) might be better?

Reply:
I can see reasonable arguments for all the possibilities, so if you feel strongly about one of them then I would be fine with it.  For example, `list()` and `describe()` even though they return the same thing (now -- `describe()` could return more later potentially).  Just let me know.

Reply:
I thought about this more and I think I see a good way out of this difficulty.  We should just have an accessor method like `userInfo(String name)` that returns a `KafkaFuture<user info>`.  We can dynamically create this future if needed.  Then we can have a single RPC and a single API which is just `describe`

Reply:
Ok, I think I follow now.  You are saying that we could potentially implement describe by invoking 1+N requests to Kafka: one to get the list of credentials (either the list of all of them if we are asking for them all, or the explicitly requested ones we wanted), and then another N requests to get the data for each one.  This on the surface seems like an anti-pattern, but it is not unreasonable for the case where the data is expensive to get in the first place — maybe we are forced to make 1 or more round-trips for each anyway.  So as a general, reusable pattern, yes, I believe it works.

So when we invoke describe, whether it is describe-all or just an explicit few, we return a single future, and that future returns a list of instances (UserName in this case): either 1 instance for each user that has at least 1 credential for the describe-all case, or one instance per distinct user explicitly requested otherwise.  Then each UserName instance has the accessor you mentioned, which in this case returns  Future<UserScramCredentialDescription>.

Do I have that right?

Reply:
Ok, we have two levels of Futures now:

1. `DescribeUserScramCredentialsResult` has a `KafkaFuture<List<UserScramCredentialsDescriptionResult>>`
2. `UserScramCredentialsDescriptionResult` has a user name and a `KafkaFuture<UserScramCredentialsDescription>`

I think this is where we want to end up.  Let me know if you agree.  I also added some logic in `AlterUserScramCredentialsRequestTest` to confirm the behavior of the per-user futures and associated errors.